### ISSUE TITLE
predict.bsts does not capture holiday information

### ISSUE THREAD

**Issue by tracek** [steve-the-bayesian/BOOM#25]
We're having some problems with capturing holiday information via `bsts`. The problem seems to be nearly identical to what is presented in [this stats.exchange post](https://stats.stackexchange.com/questions/355495/bsts-forecasts-not-incorporating-holiday-effects). Simon has opened a [new one](https://stats.stackexchange.com/questions/395522/predict-bsts-does-not-capture-holiday-information), in which he provided a reproducible example. 

Is there a bug in `bsts` or we run `predict` incorrectly?

**Comment by steve-the-bayesian** [comment]
The following code replicates the issue without the tidyverse stuff.


library(bsts)
set.seed(12345)

trend <- cumsum(rnorm(1095, 0, .1))
dates <- seq.Date(from = as.Date("2014-01-01"), length = length(trend), by = "day")
y <- zoo(trend + rnorm(length(trend), 0, .2), dates)

AddHolidayEffect <- function(y, dates, effect) {
  ## Adds a holiday effect to simulated data.
  ## Args:
  ##   y: A zoo time series, with Dates for indices.
  ##   dates: The dates of the holidays.
  ##   effect: A vector of holiday effects of odd length.  The central effect is
  ##     the main holiday, with a symmetric influence window on either side.
  ## Returns:
  ##   y, with the holiday effects added.
  time <- dates - (length(effect) - 1) / 2
  for (i in 1:length(effect)) {
    y[time] <- y[time] + effect[i]
    time <- time + 1
  }
  return(y)
}

## Define some holidays.
memorial.day <- NamedHoliday("MemorialDay")
memorial.day.effect <- c(-.75, -2, -2)
memorial.day.dates <- as.Date(c("2014-05-26", "2015-05-25", "2016-05-25"))
y <- AddHolidayEffect(y, memorial.day.dates, memorial.day.effect)

## The holidays can be in any order.
holiday.list <- list(memorial.day)

## Let's train the model to just before MemorialDay
cut.date = as.Date("2016-05-20")
train.data <- y[time(y) < cut.date]
test.data <- y[time(y) >= cut.date]
ss <- AddLocalLevel(list(), train.data)
ss <- AddRegressionHoliday(ss, train.data, holiday.list = holiday.list)
model <- bsts(train.data, state.specification = ss, niter = 500, ping = 0)
## Now let's make a prediction covering MemorialDay
my.horizon = 15
## Note adding the time stamps here doesn't help either
pred <- predict(object = model, horizon = my.horizon)
## Make a data frame for plotting
plot(pred, plot.original = 365)

**Comment by steve-the-bayesian** [comment]
printf debugging confirms that the 'observation_matrix' method in regression_holiday_state_model is firing, but it is not hitting the holiday.

**Comment by steve-the-bayesian** [comment]
The issue was a call to "observe_time_dimension", which was done correctly when training.  When forecasting, the value passed to observe_time_dimension was less than the actual time_timension + forecast horizon, so the holiday model reported an effect of zero.

I added an error message when holidays see a time dimension that is too large, and I fixed the function call in model_manager.cc.  Parallel changes will be needed in dirm and mbsts.
